dev/MANUALLY_GENERATING_IMAGE_CACHE_AND_CONSTRAINTS.md

```diff
@@ -532,3 +359,196 @@ breeze release-management update-constraints --constraints-repo /home/user/airfl
+to be found compatible with each other. In case new dependencies are released, sometimes `pip` enters
+a long loop trying to figure out if the newly released dependency can be used, but due to some other
+dependencies of ours it is impossible, but it will take `pip` a very long time to figure it out.
+
+This is visible in the "build output" as `pip` attempting to continuously backtrack and download many new
+versions of various dependencies, trying to find a good match.
+
+This is why we sometimes we need to help pip to skip newer versions of those dependencies, until the
+condition that caused the backtracking is solved.
+
+We do it by adding `dependency<=version` to the EAGER_UPGRADE_ADDITIONAL_REQUIREMENTS variable in
+`Dockerfile.ci`. The trick is to find the dependency that is causing the backtracking.
+
+Here is how. We use `bisecting` methodology to try out candidates for backtrack triggering among the
+candidates that have been released in PyPI since the last time we successfully run
+``--upgrade-to-newer-dependencies`` and committed the constraints in the `canary` build.
+
+## How to figure out backtracking dependencies
+
+First - we have a breeze command that can help us with that:
+
+```bash
+breeze ci find-backtracking-candidates
+```
+
+This command should be run rather quickly after we notice that the CI build is taking a long time and fail,
+because it is based on the fact that eager upgrade produced valid constraints at some point of time and
+it tries to find out what dependencies have been added since then and limit them to the version that
+was used in the constraints.
+
+You can also - instead of running the command manually rely on the failing CI builds. We run the
+`find-backtracking-candidates` command in the `canary` build when it times out, so the
+easiest way to find backtracking candidates is to find the first build that failed with timeout - it
+will likely have the smallest number of backtracking candidates. The command outputs the limitation
+for those backtracking candidates that are guaranteed to work (because they are taken from the latest
+constraints and they already succeeded in the past when the constraints were updated).
+
+Then we run ``breeze ci-image build --upgrade-to-newer-dependencies --eager-upgrade-additional-requirements "REQUIREMENTS"``
+to check which of the candidates causes the long builds. Initially you put there the whole list of
+candidates that you got from the `find-backtracking-candidates` command. This **should** succeed. Now,
+the next step is to narrow down the list of candidates to the one that is causing the backtracking.
+
+We narrow-down the list by "bisecting" the list. We remove half of the dependency limits and see if it
+still works or not. It works - we continue. If it does not work, we restore the removed half and remove
+the other half. Rinse and repeat until there is only one dependency left - hopefully
+(sometimes you will need to leave few of them).
```

Comment:
You made me laugh real hard with "Rinse and repeat", but I don't think that it fits this kind of docs :)

```suggestion
still works or not. If it works, we continue. Otherwise, we restore the removed half and remove
the other half. Repeat until there is only one dependency left - hopefully
(sometimes you will need to leave few of them).
```